pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,628 @@
+		if nodeSelector.Matches(labels.Set(node.GetLabels())) {
+			pools.Insert(ipPool.Name)
+		}
+	}
+	return pools
+}
+
+func (c *Cluster) memberNum() int {
+	c.memberRWLock.RLock()
+	defer c.memberRWLock.RUnlock()
+	return c.existingMembers.Len()
+}
+
+func (c *Cluster) addExistingMembers(members ...string) {
+	c.memberRWLock.Lock()
+	defer c.memberRWLock.Unlock()
+	c.existingMembers.Insert(members...)
+	klog.V(2).InfoS("Add existingMember", "memberNum", c.existingMembers.Len())
+}
+
+func (c *Cluster) updateExistingMember(node *corev1.Node, add bool) {
+	c.memberRWLock.Lock()
+	defer c.memberRWLock.Unlock()
+	nodeAddr, err := k8s.GetNodeAddr(node)
+	if err != nil {
+		klog.ErrorS(err, "Failed to obtain local IP address")
+		return
+	}
+	member := fmt.Sprintf("%s:%d", nodeAddr, c.bindPort)
+	oldNum := c.existingMembers.Len()
+	if add {
+		c.existingMembers.Insert(member)
+		newNum := c.existingMembers.Len()
+		klog.V(2).Infof("Add new member (%s) in local existing memberlist, num %d->%d", member, oldNum, newNum)
+	} else {
+		if !c.existingMembers.Has(member) {
+			return
+		}
+		c.existingMembers.Delete(member)
+		newNum := c.existingMembers.Len()
+		klog.V(2).Infof("Delete member (%s) in local existing memberlist, num %d->%d", member, oldNum, newNum)
+	}
+	if c.nodeName == node.Name {
+		c.updateLocalNodeStatus()
+	}
+}
+
+func (c *Cluster) joinOrLeaveCluster(join bool) {
+	if join {
+		c.memberRWLock.RLock()
+		defer c.memberRWLock.RUnlock()
+		clusterNodes := c.existingMembers.List()
+		n, err := c.mList.Join(clusterNodes)
+		if err != nil {
+			klog.ErrorS(err, "Join cluster failed, cluster nodes %v", clusterNodes)
+			return
+		}
+		klog.V(2).InfoS("Join cluster", "num", n, "cluster nodes", clusterNodes)
+	} else {
+		if err := c.localNodeLeave(); err != nil {
+			klog.ErrorS(err, "Leave cluster failed")
+			return
+		}
+		klog.V(2).InfoS("Leave cluster", "nodeName", c.nodeName)
+	}
+}
+
+func (c *Cluster) localNodeLeave() error {
+	err := c.mList.Leave(defaultLeavePeriod)
+	if err != nil {
+		klog.ErrorS(err, "Local node leave cluster failed")
+		return err
+	}
+	if err := c.mList.Shutdown(); err != nil {
+		klog.ErrorS(err, "Shutdown cluster failed")
+		return err
+	}
+	newML, err := memberlist.Create(c.mlConfig)
+	if err != nil {
+		klog.ErrorS(err, "Create cluster failed")
+		return err
+	}
+	c.mList = newML
+	return nil
+}
+
+func (c *Cluster) Run(stopCh <-chan struct{}) {
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.ipPoolInformerHasSynced, c.nodeListerSynced) {
+		return
+	}
+
+	defer func() {
+		if err := c.mList.Shutdown(); err != nil {
+			klog.ErrorS(err, "Shut down cluster failed")
+		}
+		close(c.nodeEventsCh)
+	}()
+
+	for {
+		select {
+		case <-stopCh:
+			return
+		case nodeEvent := <-c.nodeEventsCh:
+			c.syncConsistentHashMap(&nodeEvent)
+		case <-c.localNodeEventCh:
+			c.syncLocalNodeStatus()
+		case hashEvent := <-c.conHashEventCh:
+			c.syncConHash(&hashEvent)
```

Comment:
what's the difference between syncConsistentHashMap and syncConHash? hard to tell from the names.

Reply:
c.syncConsistentHashMap(&nodeEvent), syncConsistentHashMap handle events from cluster, when a node join or leave from cluster.
c.syncConHash(&hashEvent) handle events from externalIPPool informer handlers and node informer handlers.
It is needed to change the name